File: src/engine/library/commands.cpp
/*
 * niepce - library/commands.cpp
 *
 * Copyright (C) 2007-2017 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <array>

#include <boost/any.hpp>

#include "fwk/base/debug.hpp"
#include "fwk/utils/pathutils.hpp"
#include "engine/db/library.hpp"
#include "engine/db/libfolder.hpp"
#include "engine/db/libfile.hpp"
#include "engine/db/libmetadata.hpp"
#include "engine/db/filebundle.hpp"
#include "engine/db/keyword.hpp"
#include "engine/db/label.hpp"
#include "commands.hpp"

using fwk::FileList;

namespace eng {

void Commands::cmdListAllKeywords(const Library::Ptr & lib)
{
    KeywordListPtr l(new KeywordList);
    lib->getAllKeywords(l);
    /////
    // notify folder added l
    lib->notify(LibNotification::make<LibNotification::Type::ADDED_KEYWORDS>({l}));
}

void Commands::cmdListAllFolders(const Library::Ptr & lib)
{
    LibFolderListPtr l(new LibFolderList);
    lib->getAllFolders(l);
    /////
    // notify folder added l
    lib->notify(LibNotification::make<LibNotification::Type::ADDED_FOLDERS>({l}));
}

void Commands::cmdImportFile(const Library::Ptr & lib,
                             const std::string & file_path,
                             LibraryManaged manage)
{
    DBG_ASSERT(manage == LibraryManaged::NO,
               "managing file is currently unsupported");

    FileBundlePtr bundle(filebundle_new());
    engine_db_filebundle_add(bundle.get(), file_path.c_str());

    std::string folder = fwk::path_dirname(file_path);

    LibFolderPtr pf = lib->getFolder(folder);
    if(!pf) {
        pf = lib->addFolder(folder);
        LibFolderListPtr l(new LibFolderList);
        l->push_back(pf);
        lib->notify(LibNotification::make<LibNotification::Type::ADDED_FOLDERS>({l}));
    }
    lib->addBundle(engine_db_libfolder_id(pf.get()), bundle, manage);
    lib->notify(LibNotification::make<LibNotification::Type::ADDED_FILES>({}));
}

void Commands::cmdImportFiles(const Library::Ptr & lib,
                              const std::string & folder,
                              const FileList::Ptr & files, LibraryManaged manage)
{
    DBG_ASSERT(manage == LibraryManaged::NO,
               "managing file is currently unsupported");

    FileBundleListPtr bundles = filebundle_filter_bundles(files);

    LibFolderPtr pf = lib->getFolder(folder);
    if(!pf) {
        pf = lib->addFolder(folder);
        LibFolderListPtr l(new LibFolderList);
        l->push_back(pf);
        lib->notify(LibNotification::make<LibNotification::Type::ADDED_FOLDERS>({l}));
    }
    std::for_each(bundles->begin(), bundles->end(),
                  [&lib, &pf, manage] (const auto& fb) {
                      lib->addBundle(engine_db_libfolder_id(pf.get()), fb, manage);
                  });
    lib->notify(LibNotification::make<LibNotification::Type::ADDED_FILES>({}));
}


void Commands::cmdQueryFolderContent(const Library::Ptr & lib,
                                     library_id_t folder_id)
{
    LibFileListPtr fl(new LibFileList);
    lib->getFolderContent(folder_id, fl);
    lib->notify(LibNotification::make<LibNotification::Type::FOLDER_CONTENT_QUERIED>(
                    {folder_id, fl}));
}

void Commands::cmdCountFolder(const Library::Ptr & lib,
                              eng::library_id_t folder_id)
{
    int count = lib->countFolder(folder_id);
    lib->notify(LibNotification::make<LibNotification::Type::FOLDER_COUNTED>(
                    {folder_id, count}));
}

void Commands::cmdQueryKeywordContent(const Library::Ptr & lib,
                                      library_id_t keyword_id)
{
    LibFileListPtr fl(new LibFileList);
    lib->getKeywordContent(keyword_id, fl);
    lib->notify(LibNotification::make<LibNotification::Type::KEYWORD_CONTENT_QUERIED>(
                    {keyword_id, fl}));
}

void Commands::cmdRequestMetadata(const Library::Ptr & lib,
                                  library_id_t file_id)
{
    LibMetadata::Ptr lm(new LibMetadata(file_id));
    lib->getMetaData(file_id, lm);
    lib->notify(LibNotification::make<LibNotification::Type::METADATA_QUERIED>(
                    {file_id, lm}));
}

void Commands::cmdSetMetadata(const Library::Ptr & lib,
                              library_id_t file_id, fwk::PropertyIndex meta,
                              const fwk::PropertyValue & value)
{
    metadata_desc_t m;
    m.id = file_id;
    m.meta = meta;
    m.value = value;
    lib->setMetaData(file_id, meta, value);
    lib->notify(LibNotification::make<LibNotification::Type::METADATA_CHANGED>(
                    std::move(m)));
}

void Commands::cmdWriteMetadata(const Library::Ptr & lib,
                                 library_id_t file_id)
{
    lib->writeMetaData(file_id);
}

void Commands::cmdMoveFileToFolder(const Library::Ptr & lib,
                                   library_id_t file_id, library_id_t from_folder_id,
                                   library_id_t to_folder_id)
{
    if(lib->moveFileToFolder(file_id, to_folder_id)) {
        lib->notify(LibNotification::make<LibNotification::Type::FILE_MOVED>(
                        {file_id, from_folder_id, to_folder_id}));
        lib->notify(LibNotification::make<LibNotification::Type::FOLDER_COUNT_CHANGE>(
                        {from_folder_id, -1}));
        lib->notify(LibNotification::make<LibNotification::Type::FOLDER_COUNT_CHANGE>(
                        {to_folder_id, 1}));
    }
}

void Commands::cmdListAllLabels(const Library::Ptr & lib)
{
    eng::Label::ListPtr l(new eng::Label::List);
    lib->getAllLabels(l);
    lib->notify(LibNotification::make<LibNotification::Type::ADDED_LABELS>({l}));
}

void Commands::cmdCreateLabel(const Library::Ptr & lib,
                              const std::string & s, const std::string & color)
{
    int64_t id = lib->addLabel(s, color);
    if(id != -1) {
        eng::Label::ListPtr l(new eng::Label::List);
        l->push_back(eng::Label::Ptr(new eng::Label(id, s, color)));
        lib->notify(LibNotification::make<LibNotification::Type::ADDED_LABELS>({l}));
    }
}


void Commands::cmdDeleteLabel(const Library::Ptr & lib,
                              int label_id)
{
    lib->deleteLabel(label_id);
    lib->notify(LibNotification::make<LibNotification::Type::LABEL_DELETED>({label_id}));
}


void Commands::cmdUpdateLabel(const Library::Ptr & lib,
                              eng::library_id_t label_id, const std::string & name,
                              const std::string & color)
{
    lib->updateLabel(label_id, name, color);
    eng::Label::Ptr label(new eng::Label(label_id, name, color));
    lib->notify(LibNotification::make<LibNotification::Type::LABEL_CHANGED>({label}));
}


void Commands::cmdProcessXmpUpdateQueue(const Library::Ptr & lib, bool write_xmp)
{
    lib->processXmpUpdateQueue(write_xmp);
}

}

/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:99
  End:
*/

File: src/engine/library/commands.rs
/*
 * niepce - engine/library/commands.rs
 *
 * Copyright (C) 2017-2019 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::os::raw::c_void;

use npc_fwk::PropertyValue;
use engine::db::LibraryId;
use engine::db::library;
use engine::db::library::{
    Library,
    Managed
};
use engine::db::filebundle::FileBundle;
use engine::db::keyword::Keyword;
use engine::db::label::Label;
use engine::db::libfolder::LibFolder;
use super::notification::LibNotification;
use super::notification::{
    Content,
    Count,
    FileMove,
    MetadataChange,
};
use root::eng::NiepceProperties as Np;

pub fn cmd_list_all_keywords(lib: &Library) -> bool {
    match lib.get_all_keywords() {
        Ok(list) => {
            // XXX change this to "LoadKeywords"
            for kw in list {
                if let Err(err) = lib.notify(LibNotification::AddedKeyword(kw)) {
                    err_out!("Failed to notify AddedKeyword {:?}", err);
                    return false;
                }
            }
            true
        },
        Err(err) => {
            err_out_line!("get all keywords failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_list_all_folders(lib: &Library) -> bool {
    match lib.get_all_folders() {
        Ok(list) => {
            // XXX change this to "LoadedFolders"
            for folder in list {
                if let Err(err) = lib.notify(LibNotification::AddedFolder(folder)) {
                    err_out!("Failed to notify AddedFolder {:?}", err);
                    return false;
                }
            }
            true
        },
        Err(err) => {
            err_out_line!("get_all_folders failed: {:?}", err);
            false
        }
    }
}

//
// Get the folder for import. Create it if needed otherwise return the one that exists
//
fn get_folder_for_import(lib: &Library, folder: &str) -> library::Result<LibFolder> {
    match lib.get_folder(folder) {
        Ok(lf) => Ok(lf),
        Err(library::Error::NotFound) => {
            // folder doesn't exist, we'll create it
            if let Some(name) = Library::leaf_name_for_pathname(folder) {
                match lib.add_folder(&name, Some(String::from(folder))) {
                    Ok(lf) =>  {
                        let libfolder = lf.clone();
                        if lib.notify(LibNotification::AddedFolder(lf)).is_err() {
                            err_out!("Failed to notify AddedFolder");
                        }
                        Ok(libfolder)
                    },
                    Err(err) => {
                        err_out_line!("Add folder failed {:?}", err);
                        Err(err)
                    }
                }
            } else {
                err_out_line!("Can't get folder name.");
                Err(library::Error::InvalidResult)
            }
        },
        Err(err) => {
            err_out_line!("get folder failed: {:?}", err);
            Err(err)
        }
    }
}

pub fn cmd_import_files(lib: &Library, folder: &str, files: &[String],
                        manage: Managed) -> bool {
    dbg_assert!(manage == Managed::NO, "managing file is currently unsupported");

    let bundles = FileBundle::filter_bundles(files);
    match get_folder_for_import(lib, folder) {
        Ok(libfolder) =>  {
            let folder_id = libfolder.id();
            for bundle in bundles {
                // XXX properle handle this error. Should be a failure.
                if let Err(err) = lib.add_bundle(folder_id, &bundle, manage) {
                    err_out!("Add bundle failed: {:?}", err);
                }
            }
            if lib.notify(LibNotification::AddedFiles).is_err() {
                err_out!("Failed to notify AddedFiles");
            }
            true
        },
        Err(err) => {
            err_out_line!("Get folder for import {:?}", err);
            false
        }
    }
}

pub fn cmd_create_folder(lib: &Library, name: &str, path: Option<String>) -> LibraryId {
    match lib.add_folder(name, path) {
        Ok(lf) =>  {
            let id = lf.id();
            if lib.notify(LibNotification::AddedFolder(lf)).is_err() {
                err_out!("Failed to notifu AddedFolder");
            }
            id
        },
        Err(err) =>  {
            err_out_line!("Folder creation failed {:?}", err);
            -1
        }
    }
}

pub fn cmd_delete_folder(lib: &Library, id: LibraryId) -> bool {
    match lib.delete_folder(id) {
        Ok(_) => {
            if lib.notify(LibNotification::FolderDeleted(id)).is_err() {
                err_out!("Failed to notify FolderDeleted");
            }
            true
        },
        Err(err) => {
            err_out_line!("Delete folder failed {:?}", err);
            false
        }
    }
}

pub fn cmd_request_metadata(lib: &Library, file_id: LibraryId) -> bool {
    match lib.get_metadata(file_id) {
        Ok(lm) => {
            // This time it's a fatal error since the purpose of this comand
            // is to retrieve.
            match lib.notify(LibNotification::MetadataQueried(lm)) {
                Err(err) => {
                    err_out!("Failed to notify Metadata {:?}", err);
                    false
                },
                Ok(_) => true
            }
        },
        Err(err) => {
            err_out_line!("Get metadata failed {:?}", err);
            false
        }
    }
}

pub fn cmd_query_folder_content(lib: &Library, folder_id: LibraryId) -> bool {
    match lib.get_folder_content(folder_id) {
        Ok(fl) => {
            let mut value =
                LibNotification::FolderContentQueried(unsafe { Content::new(folder_id) });
            if let LibNotification::FolderContentQueried(ref mut content) = value {
                for f in fl {
                    unsafe { content.push(Box::into_raw(Box::new(f)) as *mut c_void) };
                }
            }
            // This time it's a fatal error since the purpose of this comand
            // is to retrieve.
            match lib.notify(value) {
                Err(err) => {
                    err_out!("Failed to notify FolderContent {:?}", err);
                    false
                },
                Ok(_) => true
            }
        },
        Err(err) => {
            err_out_line!("Get folder content failed {:?}", err);
            false
        }
    }
}

pub fn cmd_set_metadata(lib: &Library, id: LibraryId, meta: Np,
                        value: &PropertyValue) -> bool {
    match lib.set_metadata(id, meta, value) {
        Ok(_) => {
            if lib.notify(LibNotification::MetadataChanged(
                MetadataChange::new(id, meta as u32, value.clone()))).is_err() {
                err_out!("Failed to notify MetadataChange");
            }
            true
        },
        Err(err) => {
            err_out_line!("set_matadata failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_count_folder(lib: &Library, id: LibraryId) -> bool {
    match lib.count_folder(id) {
        Ok(count) => {
            // This time it's a fatal error since the purpose of this comand
            // is to retrieve.
            match lib.notify(LibNotification::FolderCounted(Count{id, count})) {
                Err(err) => {
                    err_out!("Failed to notify FolderCounted {:?}", err);
                    false
                },
                Ok(_) => {
                    true
                }
            }
        },
        Err(err) => {
            err_out_line!("count_folder failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_add_keyword(lib: &Library, keyword: &str) -> LibraryId {
    match lib.make_keyword(keyword) {
        Ok(id) => {
            if lib.notify(LibNotification::AddedKeyword(Keyword::new(id, keyword))).is_err() {
                err_out!("Failed to notify AddedKeyword");
            }
            id
        },
        Err(err) => {
            err_out_line!("make_keyword failed: {:?}", err);
            -1
        }
    }
}

pub fn cmd_query_keyword_content(lib: &Library, keyword_id: LibraryId) -> bool {
    match lib.get_keyword_content(keyword_id) {
        Ok(fl) => {
            let mut content = unsafe { Content::new(keyword_id) };
            for f in fl {
                unsafe { content.push(Box::into_raw(Box::new(f)) as *mut c_void) };
            }
            // This time it's a fatal error since the purpose of this comand
            // is to retrieve.
            match lib.notify(LibNotification::KeywordContentQueried(content)) {
                Err(err) => {
                    err_out!("Failed to notify KeywordContentQueried {:?}", err);
                    false
                },
                Ok(_) => true
            }
        },
        Err(err) => {
            err_out_line!("get_keyword_content failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_count_keyword(lib: &Library, id: LibraryId) -> bool {
    match lib.count_keyword(id) {
        Ok(count) => {
            // This time it's a fatal error since the purpose of this comand
            // is to retrieve.
            match lib.notify(LibNotification::KeywordCounted(Count{id, count})) {
                Err(err) => {
                    err_out!("Failed to notify KeywordCounted {:?}", err);
                    false
                },
                Ok(_) => true
            }
        },
        Err(err) => {
            err_out_line!("count_keyword failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_write_metadata(lib: &Library, file_id: LibraryId) -> bool {
    match lib.write_metadata(file_id) {
        Ok(_) => true,
        Err(err) => {
            err_out_line!("write_metadata failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_move_file_to_folder(lib: &Library, file_id: LibraryId, from: LibraryId,
                               to: LibraryId) -> bool {
    match lib.move_file_to_folder(file_id, to) {
        Ok(_) => {
            if lib.notify(LibNotification::FileMoved(FileMove{file: file_id, from, to}))
                .is_err() {
                    err_out!("Failed to notify FileMoved");
                }
            if lib.notify(LibNotification::FolderCountChanged(Count{id: from, count: -1}))
                .is_err() {
                    err_out!("Failed to notify FileMoved");
                }
            if lib.notify(LibNotification::FolderCountChanged(Count{id: to, count: 1}))
                .is_err() {
                    err_out!("Failed to notify FileMoved");
                }
            true
        },
        Err(err) => {
            err_out_line!("move file to folder failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_list_all_labels(lib: &Library) -> bool {
    match lib.get_all_labels() {
        Ok(l) => {
            // XXX change this notification type
            for label in l {
                if let Err(err) = lib.notify(LibNotification::AddedLabel(label)) {
                    err_out!("Failed to notify AddedLabel {:?}", err);
                    return false;
                }
            }
            true
        },
        Err(err) => {
            err_out_line!("get_all_labels failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_create_label(lib: &Library, name: &str, colour: &str) -> LibraryId {
    match lib.add_label(name, colour) {
        Ok(id) => {
            let l = Label::new(id, name, colour);
            if lib.notify(LibNotification::AddedLabel(l)).is_err() {
                err_out!("Failed to notify AddedLabel");
            }
            id
        },
        Err(err) => {
            err_out_line!("add_label failed: {:?}", err);
            -1
        }
    }
}

pub fn cmd_delete_label(lib: &Library, label_id: LibraryId) -> bool {
    match lib.delete_label(label_id) {
        Ok(_) => {
            if lib.notify(LibNotification::LabelDeleted(label_id)).is_err() {
                err_out!("Failed to notify LabelDeleted");
            }
            true
        },
        Err(err) => {
            err_out_line!("delete label failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_update_label(lib: &Library, label_id: LibraryId, name: &str,
                        colour: &str) -> bool {
    match lib.update_label(label_id, name, colour) {
        Ok(_) => {
            let label = Label::new(label_id, name, colour);
            if lib.notify(LibNotification::LabelChanged(label)).is_err() {
                err_out!("Failed to notify LabelChanged");
            }
            true
        },
        Err(err) => {
            err_out_line!("update label failed: {:?}", err);
            false
        }
    }
}

pub fn cmd_process_xmp_update_queue(lib: &Library, write_xmp: bool) -> bool {
    match lib.process_xmp_update_queue(write_xmp) {
        Ok(_) => true,
        Err(err) => {
            err_out_line!("process_xmp_update_queue failed: {:?}", err);
            false
        }
    }
}
